src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -0,0 +1,345 @@
+          std::static_pointer_cast<const StoredTableNode>(column_candidate->original_node.lock());
+
+      if (expression_table == candidate_table) {
+        // both columns same table -> if both are UCC we could still convert join -> both are UCC candidate
+        candidates->push_back(UCCCandidate{expression_table->table_name, column_expression->original_column_id});
+        candidates->push_back(UCCCandidate{candidate_table->table_name, column_candidate->original_column_id});
+      }
+
+      return LQPVisitation::VisitInputs;
+    });
+  }
+
+  return candidates;
+}
+
+UCCCandidates UccDiscoveryPlugin::identify_ucc_candidates() const {
+  // get a snapshot of the current LQP cache to work on all currently cached queries
+  const auto snapshot = Hyrise::get().default_lqp_cache->snapshot();
+
+  auto ucc_candidates = UCCCandidates{};
+
+  for (const auto& [query, entry] : snapshot) {
+    const auto& root_node = entry.value;
+
+    visit_lqp(root_node, [&](auto& node) {
+      const auto type = node->type;
+      if ((type != LQPNodeType::Join) && (type != LQPNodeType::Aggregate)) {
+        // Non-Join and Non-Aggregate (Groupby) nodes are not considered for optimization using UCCs
+        return LQPVisitation::VisitInputs;
+      }
+
+      if (type == LQPNodeType::Aggregate) {
+        // in case of aggregate, extract all predicates used in groupby operations
+        auto& aggregate_node = static_cast<AggregateNode&>(*node);
+        auto column_candidates = std::vector<std::shared_ptr<AbstractExpression>>{
+            aggregate_node.node_expressions.begin(),
+            aggregate_node.node_expressions.begin() + aggregate_node.aggregate_expressions_begin_idx};
+
+        for (const auto& column_candidate : column_candidates) {
+          const auto casted_candidate = std::dynamic_pointer_cast<LQPColumnExpression>(column_candidate);
+          if (!casted_candidate) {
+            continue;
+          }
+          // every ColumnExpression used as a GroupBy expression should be checked for uniqueness
+          const auto table = std::static_pointer_cast<const StoredTableNode>(casted_candidate->original_node.lock());
+          ucc_candidates.insert(UCCCandidate{table->table_name, casted_candidate->original_column_id});
+        }
+
+        return LQPVisitation::VisitInputs;
+      }
+
+      // if not aggregate node, must be a join node
+      auto& join_node = static_cast<JoinNode&>(*node);
+      const auto& join_mode = join_node.join_mode;
+      // get join predicate with equals condition, that's the only one we would want to work on
+      std::shared_ptr<BinaryPredicateExpression> join_predicate = nullptr;
+      for (auto predicate : join_node.join_predicates()) {
+        join_predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate);
+        if (!join_predicate) {
+          continue;
+        }
+        if (join_predicate->predicate_condition != PredicateCondition::Equals) {
```

Comment:
It should be possible to do that, however, we didn't work out the details of those rewrites. It may be a future expansion on the plugin and rewrite rule though.

It should be fine to remove this condition for the UCC discovery. However, considering the time left as well as the fact that we wouldn't utilize the UCCs discovered in these new cases, I'd keep it this way to avoid some nasty side-effects we didn't think of when dealing with predicates that do not have the equals condition.

Reply:
Can you add a comment or mark this as TODO? Totally fine that this is out of scope when you document it.